controllers/noderefutil/providerid.go

```diff
@@ -100,3 +101,11 @@ func (p *ProviderID) String() string {
 func (p *ProviderID) Validate() bool {
 	return p.CloudProvider() != "" && p.ID() != ""
 }
+
+// IndexKey returns a string concatenating the cloudProvider and the ID parts of the providerID.
+// E.g Format: cloudProvider://optional/segments/etc/id. IndexKey: cloudProvider/id
+// This is useful to use the providerID as a reliable index between nodes and machines
+// as it guarantees the infra Providers contract.
+func (p *ProviderID) IndexKey() string {
+	return fmt.Sprintf("%s/%s", p.CloudProvider(), p.ID())
```

Comment:
I know there was a previous discussion on this but looks like it wasn't resolved, wouldn't this break with some providers? Does this also break for Azure when the vmss ID can be the same across different scale sets @alexeldeib ?

Reply:
Its already broken for VMSS, this is no more broken than the existing logic unfortunately.

I think this PR is good, it doesn’t affect the existing behavior much besides adding the index and falling back to listing. we really need to do full provider ID comparisons everywhere for a fix on the issue you mentioned.

Reply:
tracked here https://github.com/kubernetes-sigs/cluster-api/issues/4526